core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,251 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+import com.mesosphere.usi.models._
+
+object SchedulerLogic {
+  private def assertValidTransition(oldSpec: PodSpec, newSpec: PodSpec): Unit = {
+    if ((oldSpec.goal == Goal.Terminal) && (newSpec.goal == Goal.Running))
+      throw new IllegalStateException(s"Illegal state transition Terminal to Running for podId ${oldSpec.id}")
+  }
+
+  def taskIdsFor(pod: PodSpec): Seq[TaskId] = {
+    // ignoring momentarily that these might be different
+    Seq(pod.id.value)
+  }
+
+  def handlePodSpecUpdate(podSpecs: Map[PodId, PodSpec], podRecords: Map[PodId, PodRecord])
+                         (podId: PodId, newState: Option[PodSpec], responseBuilder: ResponseBuilder): ResponseBuilder = {
+    var b = responseBuilder
+
+    newState match {
+      case None =>
+        b = b.withChangeMessage(PodStatusChange(podId, None))
```

Comment:
I don't understand, why do we need two here? Why is `PodStatusChange` not enough?

Reply:
Why do we need two what here? PodStatus and PodRecord?

Because the records are updated independently. A PodStatus may not have a PodRecord, and vice versa. When we launch a pod, it has no status yet, but we must have a record that we launched it. And, if we discover a status for which there is no record, then its a spurious task.

If we coupled task status with pod record then we would needlessly need to rewrite the podRecord every time a Mesos health check was received or whatnot.